stimpacks/deploy/config.go

```diff
@@ -0,0 +1,352 @@
+	Secrets         []*v2e.SecretItem `yaml:"secrets"`
+	EnvironmentVars []*EnvironmentVar `yaml:"env"`
+}
+
+// Kubernetes describes the Kubernetes configuration to use
+type Kubernetes struct {
+	ServiceAccount string `yaml:"serviceAccount"`
+	Cluster        string `yaml:"cluster"`
+}
+
+// Environment describes a deployment environment (i.e. dev, stage, prod, etc.)
+type Environment struct {
+	Name        string      `yaml:"name"`
+	Spec        *Spec       `yaml:"spec"`
+	Instances   []*Instance `yaml:"instances"`
+	instanceMap map[string]int
+}
+
+// Instance describes an instance of a deployment within an environment (i.e. us-west-2 for env prod)
+type Instance struct {
+	Name string `yaml:"name"`
+	Spec *Spec  `yaml:"spec"`
+}
+
+// EnvironmentVar describes a shell env var to be injected into the deployment environment
+type EnvironmentVar struct {
+	Name  string `yaml:"name"`
+	Value string `yaml:"value"`
+}
+
+// parseConfig opens the deployment config file and ensures it is valid
+func (d *Deploy) parseConfig() {
+
+	d.config = Config{}
+
+	configFile := d.stim.GetConfig("deploy.file")
+
+	if configFile == "" {
+		setConfigDefault(&configFile, defaultConfigFile)
+		d.log.Debug("Deployment file not specified, using {}", defaultConfigFile)
+	}
+
+	_, err := os.Stat(configFile)
+	if err != nil && !os.IsExist(err) {
+		d.log.Fatal("No deployment config file exists at: {}", configFile)
+	}
+
+	contentstring, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		d.log.Fatal("Deployment config file could not be read: {}", err)
+	}
+
+	if ok, err := utils.IsYaml(contentstring); !ok {
+		d.log.Fatal("Deployment config file ({}) is not valid YAML: {}", configFile, err)
+	}
+
+	err = yaml.Unmarshal([]byte(contentstring), &d.config)
+	if err != nil {
+		d.log.Fatal("Error parsing deployment config {}", err)
+	}
+
+	d.config.configFilePath = configFile
+
+	d.processConfig()
+
+}
+
+// processConfig ensures that the deployment config is valid
+func (d *Deploy) processConfig() {
+
+	// Set defaults
+	setConfigDefault(&d.config.Deployment.Container.Repo, defaultContainerRepo)
+	setConfigDefault(&d.config.Deployment.Container.Tag, defaultContainerTag)
+	setConfigDefault(&d.config.Deployment.Directory, defaultDeployDirectory)
+	setConfigDefault(&d.config.Deployment.Script, defaultDeployScript)
+
+	// Create our global spec if it doesn't exist so we don't have to keep checking if it exists
+	if d.config.Global.Spec == nil {
+		d.config.Global.Spec = &Spec{}
+	}
+
+	d.config.environmentMap = make(map[string]int)
+	for i, environment := range d.config.Environments {
+
+		// Check to make sure that we don't have multiple environments with the same name
+		if _, ok := d.config.environmentMap[environment.Name]; ok {
+			d.log.Fatal("Error parsing config, duplicate environment name `{}` found", environment.Name)
+		}
+		d.config.environmentMap[environment.Name] = i
+
+		// Ensure there are instances for this environment
+		if len(environment.Instances) <= 0 {
```

Comment:
nit: but this should move up before you assign:
```golang
d.config.environmentMap[environment.Name] = i
```

Reply:
I'm not sure I understand.  What should be moved up?

Reply:
Oh, are saying to do that to avoid the assignment if there are other errors?  Is breaking up the logical steps (readability-wise) worth the savings?

Reply:
```golang
		// Check to make sure that we don't have multiple environments with the same name
		if _, ok := d.config.environmentMap[environment.Name]; ok {
			d.log.Fatal("Error parsing config, duplicate environment name `{}` found", environment.Name)
		}
		// Ensure there are instances for this environment
		if len(environment.Instances) <= 0 {
			d.log.Fatal("No instances found for environment: `{}`", environment.Name)
		}
		d.config.environmentMap[environment.Name] = i
```

basically do all possible checks before any assignment.  Granted since this exits the process completely its not a big deal but in a case where it might just return an error now the struct state might be a little messed up if something else uses it.  Also easier to read error conditions.